src/vision_main.cpp

```diff
@@ -0,0 +1,127 @@
+#include "../include/vision.h"
+// #include "constants.h"
+#include <cstdlib>
+#include "opencv2/opencv.hpp"
+#include <stdio.h>
+#include <iostream>
+#include <jsoncpp/json/json.h>
+#include <memory>
+#include <fstream>
+
+std::vector<float> jsonToFloatVector(const Json::Value& jsonArray)
+{
+  std::vector<float> result;
+  if (!jsonArray.isArray())
+  {
+    std::cerr << "Input JSON is not an array!" << std::endl;
+    return result;
+  }
+  for (unsigned int i = 0; i < jsonArray.size(); i++)
+  {
+    if (jsonArray[i].isDouble() or jsonArray[i].isInt())
+    {
+      result.push_back(static_cast<float>(jsonArray[i].asFloat()));
+    }
+    else
+    {
+      std::cerr << "Element is not a float number!" << std::endl;
+    }
+  }
+  return result;
+}
+
+class CameraCalibration
+{
+public:
+  int height;
+  int width;
+  std::string type;
+  std::vector<float> camera_matrix;
+  std::vector<float> distortion;
+  std::vector<float> rectification;
+  std::vector<float> projection;
+  CameraCalibration(const Json::Value& jsonArray)
+  {
+    if (jsonArray.isNull())
+    {
+      std::cerr << "Input JSON is Null!" << std::endl;
+    }
+    else
+    {
+      height = jsonArray["height"].asInt();
+      width = jsonArray["width"].asInt();
+
+      type = jsonArray["type"].asString();
+      camera_matrix = jsonToFloatVector(jsonArray["camera_matrix"]);
+      distortion = jsonToFloatVector(jsonArray["distortion"]);
+      rectification = jsonToFloatVector(jsonArray["rectification"]);
+      projection = jsonToFloatVector(jsonArray["projection"]);
+    }
+  }
+  ~CameraCalibration()
+  {
+  }
+};
+
+std::string keys = "{ help  h        |                       | Print help message. }"
+                   "{ config c       | ../config/config.json | Usage: Path to json file.}"
+                   "{ use_rgb        | true                  | Usage: enable use of RGB cam.}"
+                   "{ use_depth      | true                  | Usage : enable use Depth cam.}"
+                   "{ rgbresolution  | 320x240               | Available: 320x240, 640x480, 1280x720, 1920x1080}"
+                   "{ depthresolution| 424x240               | Available: 424x240, 480x270}";
+
+int main(int argc, char** argv)
+{
+  cv::CommandLineParser parser(argc, argv, keys);
+
+  parser.about("Non-ROS library for KINOVA® KORTEX™ arms vision module.");
+  if (parser.has("help"))
+  {
+    parser.printMessage();
+    return 0;
+  }
+  Json::Value param;
+  std::ifstream config_doc(parser.get<std::string>("config"));
+  config_doc >> param;
+  std::string device_ip = param["device_ip"].asString();
+  std::vector<float> static_transformation = jsonToFloatVector(param["static_transformation"]);
+  std::unique_ptr<Vision> kv(new Vision());
```

Comment:
I think we should pass the config + arguments to the `Vision` class (either in the constructor or a `configure` function), and let it do all the parsing. I see `vision_main.cpp` as an example of how to use the library, so as far as possible it should not have code that will remain the same regardless of how the user uses the library (i.e. move as much as possible into the library).

Otherwise, this looks like exactly the functionality that is needed.

Reply:
Since you had mentioned keeping the json parser separate from the library, it would still be nice to separate it out of the `main`, even if it's not part of `vision.cpp`.

Reply:
I now start to think it makes sense to use the class to makes two objects, one for calling the RGB, one for the Depth.
e.g.
```
std::unique_ptr<Vision> kv_rgb(new Vision(rgb_config,...));
std::unique_ptr<Vision> kv_depth(new Vision(depth_config...));
cv::Mat rgb_img, depth_img;
kv_rgb->run();
kv_depth->run();
kv_rgb->get_img(rgb_img);
kv_rgb->get_img(depth_img);
...
```
I think this would simplify the class significantly. what do you think?

Reply:
For point cloud generation we would ideally want to get the RGB and depth frames at the same time (which I think is possible with the single pipeline with two sources and a common clock - I'm not sure of this though). Another downside of two objects is that we will probably need a second class to perform point cloud generation. 

There's no harm in trying it out though, and refactoring later if it doesn't work.

Reply:
Willco